src/ipyradiant/visualization/interactive_exploration.py

```diff
@@ -0,0 +1,324 @@
+        """
+        This function expands a node by loading in its predicates and subjects when
+        a node is selected and the expand button is clicked.
+        """
+        self.undo_button.disabled = False
+        if self.selected_node is None:
+            return None
+        new_data = GetOutgoingPredicateObjects.run_query(
+            graph=self.rdf_graph, subject=self.selected_node.data["iri"]
+        )
+        objs = new_data["o"].tolist()
+        preds = new_data["p"].tolist()
+        labels = new_data["label"].tolist()
+        # add nodes
+        self.existing_node_ids = [
+            node.data["id"] for node in self.cyto_graph.graph.nodes
+        ]
+        self.new_nodes = {}
+        self.new_edges = {}
+        for ii, x in enumerate(objs):
+            if str(x) not in self.existing_node_ids:
+
+                self.new_nodes[ii] = Node(
+                    data={
+                        "id": str(x),
+                        "iri": x,
+                        "_label": labels[ii] or str(x),
+                    },
+                    classes="temp",
+                )
+                self.cyto_graph.graph.add_node(self.new_nodes[ii])
+            self.new_edges[ii] = Edge(
+                data={
+                    "source": self.selected_node.data["id"],
+                    "target": str(x),
+                    "iri": URItoID(preds[ii]),
+                },
+                classes="temp",
+            )
+
+            self.cyto_graph.graph.add_edge(self.new_edges[ii])
+        self.cyto_graph.set_layout(name="cola")
+
+    def undo_expansion(self, b):
+        """
+        This is a preliminary function for undoing expansions upon a node.
+        As of right now, a user can only undo the most recent expansion. After doing this,
+        the button will be disabled until a new expansion is made.
+        """
+        self.undo_button.disabled = True
+        for node in self.new_nodes:
+            self.cyto_graph.graph.remove_node_by_id(self.new_nodes[node].data["id"])
+        for edge in self.new_edges:
+            try:
+                self.cyto_graph.graph.remove_edge(self.new_edges[edge])
+            except ValueError:
+                # edge already removed from graph because the node was removed earlier.
+                pass
+
+        self.cyto_graph.set_layout(name="cola")
+
+    def remove_temp_nodes(self, b):
+        """
+        This is a basic function that cycles through the graph and removes all nodes that
+        have the 'temp' style (i.e. nodes that are not starting nodes or have not been clicked on).
+        """
+        nodes_to_remove = []
+        for node in self.cyto_graph.graph.nodes:
+            if node.classes == "temp":
+                nodes_to_remove.append(node.data["id"])
+        for node in nodes_to_remove:
+            self.cyto_graph.graph.remove_node_by_id(node)
+        # change edge color
+        for edge in self.cyto_graph.graph.edges:
```

Comment:
Is this going to be an issue in the future if we want to add other classes to the edges?

It will remove any other classes without users knowing about it...